src/modules/users-permissions/index.js

```diff
@@ -30,115 +30,137 @@ class UsersPermissions extends HasModels {
   }
 
   _createRoles() {
-    return Promise.map(_.keys(this._defaultRoles), (role) => {
-      return this.models.Role.createOrUpdate({role}, {role, systemDefined: true}).then((roleObj) => {
-        this.log.info("Created role", role)
-        roleObj.permissions = _.union(roleObj.permissions || [], this._defaultRoles[role])
-        return roleObj.save()
-      })
+    return Promise.map(_.keys(this._defaultRoles), async (role) => {
+      let roleObj = await this.models.Role.createOrUpdate({role}, {role, systemDefined: true})
+      this.log.info("Created role", role)
+      roleObj.permissions = _.union(roleObj.permissions || [], this._defaultRoles[role])
+      return roleObj.save()
     })
   }
-  
+
+
   /*
    * Register a permission
    * @param {string} name Permission name
-   * @param {function|string} [handler|route] A handler or route to wrap with a permission check
-   * @param {object} [objectParams] Mapping of req params to ObjectRoleModel identity
-   * @param {string|array} [roleName] Default role name(s) to contain this permission
+   * @param {string|array} [defaultRoles] Default role name(s) to contain this permission
+   * @param {object} [objectModel] Collection of ObjectRoleModel for object permissions, or function (objectId, user) to return the roles a user has for a given object
    */
-  allow(name, handler = null, objectParams = null, roleName = null) {
+  register(name, defaultRoles=null, objectModel=null) {
+    this.log.info("Registering permission", name, "for", defaultRoles, objectModel)
     this._permissions[name] = {
       name,
-      handler,
-      objectParams,
+      objectModel,
     }
-    if(roleName) {
-      if (!_.isArray(roleName)) {
-        roleName = [roleName]
+    if(defaultRoles) {
+      if (!_.isArray(defaultRoles)) {
+        defaultRoles = [defaultRoles]
       }
-      for (let role of roleName) {
+      for (let role of defaultRoles) {
         if (!this._defaultRoles[role]) {
           this._defaultRoles[role] = []
         }
         this._defaultRoles[role].push(name)
       }
     }
+  }
+  
+  /*
+   * Guard a handler method or route a permission
+   * @param {function|string} [handler|route] A handler or route to wrap with a permission check
+   * @param {string} name Permission name
+   * @param {object} [objectParam] URL param for object permissions object key
+   * @returns {function} Wrapped handler for method handlers
+   */
+  guard(handler, name, objectParam=null) {
```

Comment:
Yeah, splitting them makes sense. objectParam is used for both, but they're pretty clear separate code paths in registration now.